src/brogue/Math.c

```diff
@@ -98,9 +98,10 @@ u4 ranval( ranctx *x ) {
     return x->d;
 }
 
-void raninit( ranctx *x, u4 seed ) {
+void raninit( ranctx *x, uint64_t seed ) {
     u4 i;
-    x->a = 0xf1ea5eed, x->b = x->c = x->d = seed;
+    x->a = 0xf1ea5eed, x->b = x->c = x->d = (u4)seed;
+    x->c ^= (u4)(seed >> 32);
```

Comment:
So if I understand correctly, the additional seeding with the high 32 bits is done with XOR, so the seeding process is fully compatible when a 32-bit number is given. This also makes no alterations to the distribution of random numbers generated. Is that correct?

Reply:
Hang on, _does this mean that we haven't actually created any new dungeons?_ Without expanding the RNG state to 64-bits, this essentially does nothing w.r.t dungeon generation, it just creates 2^32 equivalent aliases for a given seed?

Reply:
This RNG has 32-bit output but 128-bit state 😊 It's JSF by Bob Jenkins. Bob's default implementation only takes a 32-bit seed, but other implementations accept as much as 96 bits. The patch initializes the RNG with:
```
a = constant
b = seedl
c = seedh ^ seedl
d = seedl
```
(with seedh and seedl the two halves of the 64-bit seed) which definitely yields 2^64 different starting states, hence 2^64 different dungeons. But I could seed it differently if it makes the math more obvious, for example:
```
a = constant1
b = constant2
c = seedh
d = seedl
```
In any case, there might be the opportunity to switch to another RNG with stronger guarantees (and just as simple). For instance xorshift128+ which Firefox recently switched to, or its successor xoroshiro128+. Faster than JSF, 128-bit state, 64-bit output, only 1 fixed point (JSF has 6) and the cycle length is guaranteed to be 2^128-1 (JSF guarantees nothing).

Reply:
> the seeding process is fully compatible when a 32-bit number is given.

The levels will be different because they are now seeded with 64 bits. To have the exact same dungeons before/after patch for small seeds, I would have had to keep 32-bit level seeds, but as a result there would "only" be 4 billion different maps for each level, like before.

> This also makes no alterations to the distribution of random numbers generated. Is that correct?

That is correct. If the upper 32 bits of the seed are zeros, the RNG outputs the same numbers as before.

Reply:
Ah yes, I misunderstood, I thought all of a,b,c,d were XOR'd with seedh but it's only c.

>The levels will be different because they are now seeded with 64 bits.

Hmm, ok. I know that people have liked the fact that Brogue dungeons don't hugely change between versions; as we haven't touched the architect, CE dungeons are still very similar to 1.7.5.

Can I ask, what is the motivation for expanding this seeding? We aren't really close to the limit of 32-bit time-based seeds, and 4 billion is probably more than we could all play in all our lifetimes. (Though tbf that 100M level seeding limit could have been worth expanding even at 32bit, changing dungeons anyway.)

Reply:
> people have liked the fact that Brogue dungeons don't hugely change between versions

Fixed 😊 When the seed fits in 32 bits, the dungeon is now identical to 1.7.4. Below screenshot shows the repo's very first commit vs this patch, for seed "123":

![1 7 4_vs_master](https://user-images.githubusercontent.com/69065091/94034792-a3dc0200-fd7f-11ea-9359-9ff07ee598ea.png)

> what is the motivation for expanding this seeding?

Excellent question, that deserves to be asked of every non-bugfixing patch. In my case, I like big numbers 😁

Reply:
Ok, I think I like the backwards-compat, we can get rid of it if we make any dungeon changes.

Reply:
Ok, I like the backwards compatibility, we can remove it if/when we make dungeon changes